### PR TITLE
Added pathlib support to datasets/utils.py

### PR DESCRIPTION
**Pull request by ahmadsharif1**
More work related to #8120 -- this time for all of datasets/utils.py.

### CODE REVIEW

File: torchvision/datasets/utils.py
Comment (NicolasHug) [inline torchvision/datasets/utils.py:405]: Unfortunately, forcing the return type to be a `Path` when it previously was a `str` is a BC-breaking change. It doesn't really matter for `_decompress` because it's private (leading underscore), but it's an issue here with `extract_archive` which is technically public.

(Side note: we didn't intend for any of the stuff in `datasets/utils.py` to ever be public, but that's another story).

We have 2 options here:

- always return a `str`. This is sub-ideal but simpler. And since we never intended those utils to be public anyway, it's kinda OK - we should deprecate them and make them private eventually, but that's for another day.
- always return the same type as `from_path`. This is the best UX, and requires a bit more logic.

Comment (ahmadsharif1) [reply torchvision/datasets/utils.py:405]: Great point. I have done the second option (though technically speaking that also changes behavior from the public API).

File: torchvision/datasets/utils.py
Comment (NicolasHug) [inline torchvision/datasets/utils.py:437]: Would it make more sense to call `path_or_str` within `_decompress` so that only one of these functions gets changed?

**Review by NicolasHug — APPROVED** [review]
Thanks @ahmadsharif1 , one minor comment below, I'll let you decide before merging

File: torchvision/datasets/utils.py
Comment (ahmadsharif1) [reply torchvision/datasets/utils.py:437]: One issue is that that the os.path.dirname on L406 converts pathlib.Path into str, so to_path is always str if the code goes through that if condition:

```
>>> import pathlib
>>> import os
>>> a_str="/a/b"
>>> a_path=pathlib.Path(a_str)
>>> a_str
'/a/b'
>>> a_path
PosixPath('/a/b')
>>> os.path.dirname(a_str)
'/a'
>>> os.path.dirname(a_path)
'/a'

```

So if `archive_type` is not None, we can't just return `to_path` without checking the type of `from_path`.

Your thoughts?

Comment (NicolasHug) [reply torchvision/datasets/utils.py:437]: OK good points, let's keep it as-is then

**Pull request merged**
